### ISSUE TITLE
recompilation errors go to stderr, not xmonad.errors

### ISSUE THREAD

**Issue by geekosaur** [xmonad/xmonad#403]
### Problem Description

Recompilation errors go directly to standard error instead of the log file. This means that, to find out what actually went wrong with `xmonad --recompile`, you need to run it in a terminal. Otherwise, you get an `xmessage` showing the build command but not its output, at least when using a build script (but IIRC these all use the same code path, so `stack` and `ghc` should fail the same way).

### Steps to Reproduce

Introduce a syntax error into your config and press `mod-q`

### Configuration File

```haskell
module Main (main) where

import XMonad

main :: IO ()
main = xmonad def

*
```

### Checklist

  - [x] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - I tested my configuration
    - [x] With `xmonad` version 0.17.0.9 (commit 521e835 if using git)
    - [n/a] With `xmonad-contrib` version XXX (commit XXX if using git)

**Comment by slotThe** [comment]
Are you sure about this?  We're using `withFile` [here](https://github.com/xmonad/xmonad/blob/master/src/XMonad/Core.hs#L653) and our [runProc](https://github.com/xmonad/xmonad/blob/master/src/XMonad/Core.hs#L676) would seem to me to do the correct thing with regards to redireciting stderr to xmonad.errors

**Comment by geekosaur** [comment]
It's quite reliable here: nothing useful in the error popup generated from `xmonad.errors`, I have to run it in a terminal.

**Comment by slotThe** [comment]
Mh, it works perfectly fine here, but I do have my config in its own stack project with a build file[^1].  I will try to reproduce this with just the stack/ghc setup that we support ootb.

[^1]: Although the entirety of that file is

      ``` shell
      #!/bin/sh
      cd "$XMONAD_CONFIG_DIR" && stack build --reconfigure || exit
      ln -f -T "$(stack exec -- which xmonadrc)" "$1"
      ```

**Comment by slotThe** [comment]
Okay, I've now tried this with both our stack and ghc setups and can't reproduce it at all.  Can you perhaps create a repo or something with a reproducer?

**Comment by geekosaur** [comment]
That's odd since I've reproduced it here with both my normal setup and one of my testing sandboxes. Maybe it *is* relevant that both use build scripts with cabal, though. (Neither is trivial, though, else they wouldn't be using build scripts….)

**Comment by geekosaur** [comment]
And I have a minimal reproducer, again using cabal with a build script; see https://github.com/geekosaur/xmonad.hs/tree/xmonad-buildscript-fail. Compare `terminal.out` with `xmessage.out`.

**Comment by slotThe** [comment]
Indeed, I can reproduce it with that.  It seems like the build script (or cabal?  I don't know) sends compilation errors to stdout and not stderr.  When I do `xmonad --recompile 1>xmonad.stdout` I only get the content of `xmessage.out` printed in the terminal, while the (relevant) contents of `terminal.errors` are sent to `xmonad.stdout`.

**Comment by slotThe** [comment]
@geekosaur is this something that we can fix in a cabal build file?  Is it a good idea to redirect stdout to `xmonad.errors` as well?

**Comment by geekosaur** [comment]
I'm a bit surprised it's not already, to be honest.

I could arrange for my build script and/or the cabal example build script to redirect stdout to stderr. I feel like it shouldn't be necessary, though. But this sounds like an upstream cabal issue, since it's redirecting stderr to stdout.

**Comment by slotThe** [comment]
Looks like there has been attempts https://github.com/haskell/cabal/pull/4789 and, more recently, https://github.com/haskell/cabal/pull/7838.  The latter seems to be active, so hopefully this will get fixed soon.

**Comment by jrgiacone** [comment]
Not sure what changed as well, I have xmessage installed but errors are not displayed, only shown in the xmonad.errors file

**Comment by geekosaur** [comment]
That would be a different issue: if they're making it to `xmonad.errors` then it's launching `xmessage` that is failing. Is it in `/usr/bin`? If not, you may need to arrange for xmonad to be run with an appropriate `$PATH` (which most X11 startup configurations won't do for you: they force `PATH=/bin:/usr/bin`).
